File: src/converter/nameResolvers/resolveFunctionReturnTypeName.ts
import ts from "typescript";
import {NameResolver} from "../nameResolver.js";
import {capitalize} from "../../utils/strings.js";
import {TypeScriptService, typeScriptServiceKey} from "../plugins/TypeScriptPlugin.js";

export const resolveFunctionReturnTypeName: NameResolver = (node, context) => {
    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? ((node: ts.Node) => node.parent)

    const functionNode = getParent(node)
    if (!functionNode) return null
    if (!ts.isFunctionDeclaration(functionNode)) return null
    if (functionNode.name === undefined) return null
    if (functionNode.type !== node) return null

    const parentName = functionNode.name.text

    return `${capitalize(parentName)}Result`
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/nameResolvers/resolveFunctionReturnTypeName.kt
package io.github.sgrishchenko.karakum.extension.nameResolvers

import io.github.sgrishchenko.karakum.extension.NameResolver
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.capitalize
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.Node
import typescript.isFunctionDeclaration

val resolveFunctionReturnTypeName = NameResolver<Node> { node, context ->
    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    val getParent = { it: Node ->
        typeScriptService?.getParent(it) ?: it.getParentOrNull()
    }

    val functionNode = getParent(node)
        ?: return@NameResolver null
    if (!isFunctionDeclaration(functionNode)) return@NameResolver null

    val functionNameNode = functionNode.name
        ?: return@NameResolver null
    if (functionNode.type != node) return@NameResolver null

    val parentName = functionNameNode.text

    "${capitalize(parentName)}Result"
}
